### DISCUSSION TITLE
Is there a way to have more than eight options?

### DISCUSSION THREAD

**Discussion by ToddRopog** [dotnet/command-line-api#2194] [Q&A]
Right now SetHandler only supports up to eight options. Is there a way to exceed this limit? Or will it be increased in the future?

**Comment by KalleOlaviNiemitalo** [comment]
In 2.0.0-beta4.22272.1, use an overload that gives the InvocationContext to the handler: <https://github.com/dotnet/command-line-api/blob/209b724a3c843253d3071e8348c353b297b0b8b5/src/System.CommandLine/Handler.Action.cs#L14-L20> <https://github.com/dotnet/command-line-api/blob/209b724a3c843253d3071e8348c353b297b0b8b5/src/System.CommandLine/Handler.Func.cs#L23-L29>

Then make the handler read InvocationContext.ParseResult <https://github.com/dotnet/command-line-api/blob/209b724a3c843253d3071e8348c353b297b0b8b5/src/System.CommandLine/Invocation/InvocationContext.cs#L81-L84> and query the values from the ParseResult: <https://github.com/dotnet/command-line-api/blob/209b724a3c843253d3071e8348c353b297b0b8b5/src/System.CommandLine/Parsing/ParseResult.cs#L162-L168>

In later versions of System.CommandLine (which are not yet at nuget.org), the SetHandler overloads that take IValueDescriptor\<T\> parameters have already been deleted, but the InvocationContext approach is still available (although the type isn't called InvocationContext any more).

**Reply by donRumatta** [reply]
Could you, please, give me another workaround. I have more than 8 params in handler method but several of them are injected services provided with BinderBase. I can switch to InvocationContext to reach options' values, but how can I add services to the context?

**Comment by ohads-MSFT** [comment]
Sample from docs (https://github.com/dotnet/command-line-api/issues/1750):
```cs
rootCommand.SetHandler(
    (context) => // InvocationContext, but it's inferred so you don't need to specify it.
    {
        var i = context.ParseResult.GetValueForOption(option);
        var s = context.ParseResult.GetValueForArgument(argument);
        var someDependency = context.GetService(typeof(ILogger)) as ILogger;
        /* Do something exciting! */ 
    });
```

**Comment by calacayir** [comment]
How about this? Your whole app:
```c#

Cli.Ext.ConfigureServices(services =>
{
    var loggerFactory = LoggerFactory.Create(
        builder => builder
            .AddConsole()
            .SetMinimumLevel(LogLevel.Information));

    services.AddSingleton(loggerFactory);
});


return await Cli.RunAsync<ThumbnailCommand>(args);


[CliCommand]
public class ThumbnailCommand
{
    private readonly ILoggerFactory loggerFactory;

    public ThumbnailCommand(ILoggerFactory loggerFactory)
    {
        this.loggerFactory = loggerFactory;
    }

    [CliOption]
    public TimeSpan StartTime { get; set; }

    [CliOption]
    public TimeSpan Interval { get; set; }

    [CliOption]
    public TimeSpan EndTime { get; set; }

    [CliOption]
    public FileInfo OutputFile { get; set; }

    [CliOption]
    public FileInfo InputFile { get; set; }

    [CliOption]
    public bool FastMode { get; set; }

    [CliOption]
    public FileInfo WebVTT { get; set; }

    [CliOption]
    public string ImagePath { get; set; }

    [CliOption]
    public int FrameWidth { get; set; }

    [CliOption]
    public int FrameHeight { get; set; }

    [CliOption]
    public int Columns { get; set; }

    [CliOption]
    public int Rows { get; set; }

    [CliOption]
    public int BorderWidth { get; set; }

    [CliOption]
    public string BackgroundGradientStart { get; set; }

    [CliOption]
    public string BackgroundGradientEnd { get; set; }

    public async Task<int> RunAsync()
    {
        var opts = new ThumbGenOptions()
            .WithStartTime(StartTime);

        //...

        await thumbnailGenerator.ExecuteAsync(opts);

        return 0;
    }
}
```
This is possible with [DotMake.CommandLine](https://github.com/dotmake-build/command-line) which makes System.CommandLine easy to use and maintain. Check it out!
